akka-actor-typed/src/main/scala/akka/actor/typed/Behavior.scala

```diff
@@ -396,38 +418,50 @@ object Behavior {
    * Execute the behavior with the given message
    */
   def interpretMessage[T](behavior: Behavior[T], ctx: TypedActorContext[T], msg: T): Behavior[T] =
-    interpret(behavior, ctx, msg)
+    interpret(behavior, ctx, msg, isSignal = false)
 
   /**
    * Execute the behavior with the given signal
    */
   def interpretSignal[T](behavior: Behavior[T], ctx: TypedActorContext[T], signal: Signal): Behavior[T] = {
-    val result = interpret(behavior, ctx, signal)
+    val result = interpret(behavior, ctx, signal, isSignal = true)
     // we need to throw here to allow supervision of deathpact exception
     signal match {
       case Terminated(ref) if result == UnhandledBehavior => throw DeathPactException(ref)
       case _                                              => result
     }
   }
 
-  private def interpret[T](behavior: Behavior[T], ctx: TypedActorContext[T], msg: Any): Behavior[T] = {
-    behavior match {
-      case null => throw new InvalidMessageException("[null] is not an allowed behavior")
-      case SameBehavior | UnhandledBehavior =>
+  private def interpret[T](behavior: Behavior[T],
+                           ctx: TypedActorContext[T],
+                           msg: Any,
+                           // optimization to avoid an instanceof on the message
+                           isSignal: Boolean): Behavior[T] = {
+    if (behavior eq null)
+      throw InvalidMessageException("[null] is not an allowed behavior")
+
+    (behavior._tag: @switch) match {
```

Comment:
can also be used in `ActorAdapter.next`

Reply:
BTW Behavior.wrap isn't used from anywhere

Reply:
Yeah, I tried that (and a few more places) but it didn't show up in the results, so I reverted. That was after the first step though, so maybe after polishing off other things it became more significant/worth doing.